suds/transport/twisted_transport.py

```diff
@@ -101,6 +105,182 @@ def creatorForNetloc(self, hostname, port):
+        received from the server correctly identifies the specified hostname.
+        The second purpose is (if the local C{pyOpenSSL} supports it) to use
+        the U{Server Name Indication extension
+        <https://en.wikipedia.org/wiki/Server_Name_Indication>} to indicate to
+        the server which certificate should be used.
+    @type hostname: L{unicode}
+
+    @param trustRoot: Specification of trust requirements of peers.  This may
+        be a L{Certificate} or the result of L{platformTrust}.  By default it
+        is L{platformTrust} and you probably shouldn't adjust it unless you
+        really know what you're doing.  Be aware that clients using this
+        interface I{must} verify the server; you cannot explicitly pass C{None}
+        since that just means to use L{platformTrust}.
+    @type trustRoot: L{IOpenSSLTrustRoot}
+
+    @param clientCertificate: The certificate and private key that the client
+        will use to authenticate to the server.  If unspecified, the client
+        will not authenticate.
+    @type clientCertificate: L{PrivateCertificate}
+
+    @param extraCertificateOptions: keyword-only argument; this is a dictionary
+        of additional keyword arguments to be presented to
+        L{CertificateOptions}.  Please avoid using this unless you absolutely
+        need to; any time you need to pass an option here that is a bug in this
+        interface.
+    @type extraCertificateOptions: L{dict}
+
+    @param kw: (Backwards compatibility hack to allow keyword-only arguments on
+        Python 2.  Please ignore; arbitrary keyword arguments will be errors.)
+    @type kw: L{dict}
+
+    @return: A client connection creator.
+    @rtype: L{IOpenSSLClientConnectionCreator}
+    """
+    extraCertificateOptions = kw.pop('extraCertificateOptions', None) or {}
+    if kw:
+        raise TypeError(
+            "optionsForClientTLS() got an unexpected keyword argument"
+            " '{arg}'".format(
+                arg=kw.popitem()[0]
+            )
+        )
+    if not isinstance(hostname, unicode):
+        raise TypeError(
+            "optionsForClientTLS requires text for host names, not "
+            + hostname.__class__.__name__
+        )
+    if clientCertificate:
+        extraCertificateOptions.update(
+            privateKey=clientCertificate.privateKey.original,
+            certificate=clientCertificate.original
+        )
+
+    # Only pass the trustRoot if it is not None to avoid mutually exclusive param issues.
+    if trustRoot:
+        certificateOptions = OpenSSLCertificateOptions(
+            trustRoot=trustRoot,
+            **extraCertificateOptions
+        )
+    else:
+        certificateOptions = OpenSSLCertificateOptions(**extraCertificateOptions)
+
+    return ClientTLSOptions(hostname, certificateOptions.getContext())
+
+
+@implementer(IOpenSSLClientConnectionCreator)
+class ClientTLSOptions(object):
```

Comment:
We should derive from the Twisted `ClientTLSOptions` class so that we don't have to implement `clientConnectionForTLS`. Also copying and pasting all of this documentation doesn't help much.